diffrax/_term.py

```diff
@@ -776,3 +785,136 @@ def _to_vjp(_y, _diff_args, _diff_term):
         dy, vjp = jax.vjp(_to_vjp, y, diff_args, diff_term)
         da_y, da_diff_args, da_diff_term = vjp((-(a_y**ω)).ω)
         return dy, da_y, da_diff_args, da_diff_term
+
+
+# The Underdamped Langevin SDE trajectory consists of two components: the position
+# `x` and the velocity `v`. Both of these have the same shape. So, by LangevinX we
+# denote the shape of the x component, and by LangevinTuple we denote the shape of
+# the tuple (x, v).
+_LangevinX = PyTree[Shaped[Array, "?*langevin"], "LangevinX"]
+_LangevinTuple = tuple[_LangevinX, _LangevinX]
+_LangevinBM = TypeVar("_LangevinBM", bound=Union[_LangevinX, AbstractBrownianIncrement])
+
+
+class _LangevinDiffusionTerm(AbstractTerm[_LangevinX, _LangevinBM]):
+    r"""Represents the diffusion term in the Langevin SDE:
+
+    $d \mathbf{x}_t = \mathbf{v}_t dt$
+
+    $d \mathbf{v}_t = - \gamma \mathbf{v}_t dt - u
+    \nabla f( \mathbf{x}_t ) dt + \sqrt{2 \gamma u} d W_t.$
+    """
+
+    gamma: _LangevinX
+    u: _LangevinX
+    control: AbstractBrownianPath
+
+    def vf(self, t: RealScalarLike, y: _LangevinTuple, args: Args) -> _LangevinX:
+        x, v = y
+
+        def _fun(_gamma, _u, _v):
+            return jnp.sqrt(2 * _gamma * _u) * jnp.ones(_v.shape, _v.dtype)
+
+        d_v = jtu.tree_map(_fun, self.gamma, self.u, v)
+        return d_v
+
+    def contr(self, t0: RealScalarLike, t1: RealScalarLike, **kwargs) -> _LangevinBM:
+        return self.control.evaluate(t0, t1, **kwargs)
+
+    def prod(self, vf: _LangevinX, control: _LangevinX) -> _LangevinTuple:
+        if isinstance(control, AbstractBrownianIncrement):
+            control = control.W
```

Comment:
This type annotation for `control` would imply that the `isinstance` check never triggers.

(I'm surprised some of the runtime typechecking during tests doesn't pick up on this -- if you ever pass an `AbstractBrownianIncrement` then I think it should fail to match the structure of `_LangevinX` as defined by the `vf` term.)

Morally speaking should we have this `isinstance` check, do you think? I note that we don't have it on any of the other terms -- we ask that the solvers handle this for us, rather than the term.
(No strong feelings, I'm not sure what's cleanest.)

Reply:
You're very correct! Honestly I don't remember why I put that there. Indeed I elsewhere always stick to the rule that `prod` should never be fed an `AbstractBrownianIncrement`, since unpacking that should be done by the solvers, as you say.